pkg/loop/config.go

```diff
@@ -0,0 +1,218 @@
+package loop
+
+import (
+	"context"
+	"math"
+
+	"google.golang.org/grpc"
+
+	libocr "github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	pb "github.com/smartcontractkit/chainlink-relay/pkg/loop/internal/pb"
+	"github.com/smartcontractkit/chainlink-relay/pkg/types"
+)
+
+var _ types.ConfigProvider = (*configProviderClient)(nil)
+
+type configProviderClient struct {
+	*serviceClient
+	offchainDigester libocr.OffchainConfigDigester
+	contractTracker  libocr.ContractConfigTracker
+}
+
+func newConfigProviderClient(lb *lggrBroker, cc *grpc.ClientConn) *configProviderClient {
+	c := &configProviderClient{serviceClient: newServiceClient(lb, cc)}
+	c.offchainDigester = &offchainConfigDigesterClient{pb.NewOffchainConfigDigesterClient(cc)}
+	c.contractTracker = &contractConfigTrackerClient{pb.NewContractConfigTrackerClient(cc)}
+	return c
+}
+
+func (c *configProviderClient) OffchainConfigDigester() libocr.OffchainConfigDigester {
+	return c.offchainDigester
+}
+
+func (c *configProviderClient) ContractConfigTracker() libocr.ContractConfigTracker {
+	return c.contractTracker
+}
+
+var _ libocr.OffchainConfigDigester = (*offchainConfigDigesterClient)(nil)
+
+type offchainConfigDigesterClient struct {
+	grpc pb.OffchainConfigDigesterClient
+}
+
+func (o *offchainConfigDigesterClient) ConfigDigest(config libocr.ContractConfig) (digest libocr.ConfigDigest, err error) {
+	pbCfg := pbContractConfig(config)
+	var reply *pb.ConfigDigestReply
+	reply, err = o.grpc.ConfigDigest(context.TODO(), &pb.ConfigDigestRequest{
```

Comment:
are we planning to propagate the context into libocr? (if yes, a ticket link would be nice)

Reply:
Planning to follow up here: https://smartcontract-it.atlassian.net/browse/BCF-2112
They will at least have shutdown triggers wired in. It would simplify things to have context and error wired through normally, but it may not be addressed in the initial version. We can sort of work around for now, provided that we also want to fallback to restarting plugins when they crash.